sdk/client/client.go

```diff
@@ -76,7 +87,25 @@ func (nsmc *NsmClient) ConnectRetry(ctx context.Context, name, mechanism, descri
 		name = nsmc.NscInterfaceName
 	}
 
-	outgoingMechanism, err := common.NewMechanism(cls.LOCAL, mechanism, name, description)
+	// TODO: refactor, don't hardcode, move somewhere
+	environment := map[string]string{}
+	for _, k := range os.Environ() {
+		key, value, err := ParseVariable(k)
+		if err != nil {
+			return nil, err
+		}
+		environment[key] = value
+	}
+	resourceEnvName, ok := environment["NSM_SRIOV_RESOURCE_NAME"]
+	if !ok {
+		return nil, errors.New("NSM_SRIOV_RESOURCE_NAME env variable missing")
+	}
+	pciAddress, ok := environment[resourceEnvName]
```

Comment:
I'm really curious you are able to retrieve the pci device id environment variable from client code. I've tried it for some other use case and it seemed like this env variable isn't yet to be available for init container.

Here is the snippet code (making use of client label for passing resource env name) and init container logs.
```
$ kubectl exec -ti icmp-responder-nsc-7fcd564bd-jszzx -n nsm-system env | grep PCIDEV
Defaulting container name to alpine-img.
Use 'kubectl describe pod/icmp-responder-nsc-7fcd564bd-jszzx -n nsm-system' to see all of the containers in this pod.
PCIDEVICE_INTEL_COM_INTEL_SRIOV_NETDEVICE=0000:08:02.0


$ git diff sdk/client/client.go
diff --git a/sdk/client/client.go b/sdk/client/client.go
index 8e060c5..c378d31 100644
--- a/sdk/client/client.go
+++ b/sdk/client/client.go
@@ -19,6 +19,7 @@ import (
        "context"
        "fmt"
        "io"
+       "os"
        "time"

        "github.com/pkg/errors"
@@ -45,6 +46,8 @@ const (
        ConnectionRetry = 10
        // RequestDelay - A delay between attempts, default = 5sec
        RequestDelay = time.Second * 5
+       // ResourcePoolName - resource pool name for smartvf injection
+       ResourcePoolName = "resourcepool"
)

// NsmClient is the NSM client struct
@@ -193,9 +196,23 @@ func NewNSMClient(ctx context.Context, configuration *common.NSConfiguration) (*
                configuration = &common.NSConfiguration{}
        }

+       span := spanhelper.FromContext(ctx, "Client.NewNSMClient")
+       defer span.Finish()
+
+       span.LogObject("os environ", os.Environ())
+
+       clientLabelsMap := tools.ParseKVStringToMap(configuration.ClientLabels, ",", "=")
+       span.LogObject("resource pool name", clientLabelsMap[ResourcePoolName])
+
+       if poolName, ok := clientLabelsMap[ResourcePoolName]; ok {
+               clientLabelsMap[connection.DeviceIDKey] = common.GetEnv(poolName, "device pool", false)
+               span.LogObject("device ids from the env", common.GetEnv(poolName, "device pool", false))
+               span.LogObject("clientLabelsMap", clientLabelsMap[connection.DeviceIDKey])
+       }
+
        client := &NsmClient{
                ClientNetworkService: configuration.ClientNetworkService,
-               ClientLabels:         tools.ParseKVStringToMap(configuration.ClientLabels, ",", "="),
+               ClientLabels:         clientLabelsMap,
                NscInterfaceName:     configuration.NscInterfaceName,
        }

@@ -211,3 +228,4 @@ func NewNSMClient(ctx context.Context, configuration *common.NSConfiguration) (*

        return client, nil
}
+

```
Init container logs (for above piece of code):

```
{"log":"time=\"2020-05-19T11:45:35Z\" level=info msg=\"\u003e\u003e\u003c\u003c---- os environ=[\\\"TRACER_ENAB
LED=true\\\",\\\"JAEGER_AGENT_PORT=6831\\\",\\\"KUBERNETES_PORT=tcp://10.96.0.1:443\\\",\\\"KUBERNETES_SERVICE_
PORT=443\\\",\\\"HOSTNAME=icmp-responder-nsc-7fcd564bd-jszzx\\\",\\\"SHLVL=1\\\",\\\"NSM_CLIENT_SOCKET=/var/lib/networkservicemesh/nsm.client.io.sock\\\",\\\"NSM_ADMISSION_WEBHOOK_SVC_SERVICE_PORT=443\\\",\\\"NSM_ADMISSION_WEBHOOK_SVC_PORT=tcp://10.102.123.34:443\\\",\\\"HOME=/root\\\",\\\"NSM_NAMESPACE=nsm-system\\\",\\\"WORKSPACE=/var/lib/networkservicemesh/\\\",\\\"NS_NETWORKSERVICEMESH_IO=icmp-responder?app=icmp\\\\u0026resourcepool=PCIDEVICE_INTEL_COM_INTEL_SRIOV_NETDEVICE\\\",\\\"KUBERNETES_PORT_443_TCP_ADDR=10.96.0.1\\\",\\\"PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin\\\",\\\"KUBERNETES_PORT_443_TCP_PORT=443\\\",\\\"KUBERNETES_PORT_443_TCP_PROTO=tcp\\\",\\\"NSM_ADMISSION_WEBHOOK_SVC_PORT_443_TCP_ADDR=10.102.123.34\\\",\\\"ENTRY=nsm-init\\\",\\\"NSM_ADMISSION_WEBHOOK_SVC_PORT_443_TCP_PORT=443\\\",\\\"NSM_ADMISSION_WEBHOOK_SVC_PORT_443_TCP_PROTO=tcp\\\",\\\"NSM_SERVER_SOCKET=/var/lib/networkservicemesh/nsm.server.io.sock\\\",\\\"KUBERNETES_SERVICE_PORT_HTTPS=443\\\",\\\"KUBERNETES_PORT_443_TCP=tcp://10.96.0.1:443\\\",\\\"JAEGER_AGENT_HOST=jaeger.nsm-system\\\",\\\"KUBERNETES_SERVICE_HOST=10.96.0.1\\\",\\\"PWD=/\\\",\\\"NSM_DEVICE_PLUGIN=true\\\",\\\"NSM_ADMISSION_WEBHOOK_SVC_PORT_443_TCP=tcp://10.102.123.34:443\\\",\\\"NSM_ADMISSION_WEBHOOK_SVC_SERVICE_HOST=10.102.123.34\\\"] span={}\"\n","stream":"stderr","time":"2020-05-19T11:45:35.619569469Z"}
{"log":"time=\"2020-05-19T11:45:35Z\" level=info msg=\"\u003e\u003e\u003c\u003c---- resource pool name=\\\"PCIDEVICE_INTEL_COM_INTEL_SRIOV_NETDEVICE\\\" span={}\"\n","stream":"stderr","time":"2020-05-19T11:45:35.619594169Z"}
{"log":"time=\"2020-05-19T11:45:35Z\" level=info msg=\"PCIDEVICE_INTEL_COM_INTEL_SRIOV_NETDEVICE not found.\"\n","stream":"stderr","time":"2020-05-19T11:45:35.619597622Z"}
{"log":"time=\"2020-05-19T11:45:35Z\" level=info msg=\"PCIDEVICE_INTEL_COM_INTEL_SRIOV_NETDEVICE not found.\"\n","stream":"stderr","time":"2020-05-19T11:45:35.619601547Z"}
{"log":"time=\"2020-05-19T11:45:35Z\" level=info msg=\"\u003e\u003e\u003c\u003c---- device ids from the env=\\\"\\\" span={}\"\n","stream":"stderr","time":"2020-05-19T11:45:35.619604364Z"}
{"log":"time=\"2020-05-19T11:45:35Z\" level=info msg=\"\u003e\u003e\u003c\u003c---- clientLabelsMap=\\\"\\\" span={}\"\n","stream":"stderr","time":"2020-05-19T11:45:35.619607395Z"}

```
The result of `os.Environ()` doesn't have an env variable starting PCIDEVICE, though the client pod comes up with PCIDEVICE_INTEL_COM_INTEL_SRIOV_NETDEVICE variable set by device plugin.
Am I missing something ?

Reply:
I see admission webhook mutates with pod request with adding resource requirements for init container, i missed that part. thanks.

Reply:
PCI information comes from the device plugin, so the env variable you found is the right one I believe :)

Reply:
@edwarnicke yes, it's the same approach followed in this PR, pci address comes from device plugin, device is injected into client pod through its init container whereas it's is injected directly into endpoint container.